File: codes/js/p970.js
const math = require('mathjs');

function lambertW(z, k) {
  let r = math.abs(z);
  let t = math.arg(z);
  let w = math.add(Math.log(r), math.multiply(math.i, t+2*Math.PI*k));

  for(let i=0;i<50;i++){
    let ew = math.exp(w);
    let f = math.subtract(math.multiply(w,ew),z);
    let fp = math.multiply(ew,math.add(w,1));
    w = math.subtract(w,math.divide(f,fp));
  }
  return w;
}

File: codes/java/p970.java
import org.apache.commons.math3.complex.Complex;

public class Main {
    static Complex lambertW(Complex z, int k) {
        double r = z.abs();
        double t = z.getArgument();
        Complex w = new Complex(Math.log(r), t + 2*Math.PI*k);

        for(int i=0;i<100;i++) {
            Complex ew = w.exp();
            Complex f  = w.multiply(ew).subtract(z);
            Complex fp = ew.multiply(w.add(1));
            Complex dw = f.divide(fp);
            w = w.subtract(dw);
        }
        return w;
    }
}
